libglusterfs/src/mem-pool.c

```diff
@@ -91,35 +91,28 @@ gf_mem_set_acct_info(struct mem_acct *mem_acct, struct mem_header *header,
                      size_t size, uint32_t type, const char *typestr)
 {
     struct mem_acct_rec *rec = NULL;
-    bool new_ref = false;
+    uint64_t num_allocs;
 
     if (mem_acct != NULL) {
         GF_ASSERT(type <= mem_acct->num_types);
 
         rec = &mem_acct->rec[type];
+        if (!rec->typestr) {
+            rec->typestr = typestr;
+        }
```

Comment:
I'm not sure if we really need this. This is not thread-safe (though the consequences don't seem bad) and we could get the same information by knowing the index (type) and checking the name of the memory type defined in the xlator (we could store the xlator inside mem_acct).

Reply:
I think there is no harm to setup typestr out of critical section and it would be extra work to declare a xlator object in mem_acct so we can go with same. We can get if we are getting some benefit.

Reply:
Setting the xlator in mem_acct is done once per xlator, while checking typestr is done once for each allocation and once for each free.

Note that there are two potential races here between this and code in `_gf_free()`: two threads, one allocating memory and the other releasing the last object of a given type, could execute this sequence of operations:

```
    Free                                            Allocate
    ----                                            --------
                                                    if (type.str == NULL) /* is false */
    num_allocs = atomic_dec(type.num_allocs)
    if (!num_allocs)
        type.str = NULL
        mem_acct_unref(mem_acct)
                                                    num_allocs = atomic_inc(type.num_allocs)
                                                    if (num_allocs == 1)
                                                        atomic_inc(mem_acct.refcnt)
```
After this, typestr will be NULL, but there will be allocated objects of that type. Also, if 'free' thread were releasing the last reference of mem_acct, the 'allocate' thread would probably crash. Only if we are 100% sure that all callers to `gf_mem_set_acct_info()` hold a reference to mem_acct, this can be considered ok, but the typestr problem won't be fixed.

Reply:
We can avoid set type.str only if num_allocs == 1.

Reply:
The problem is still the same:

```
    Free                                            Allocate
    ----                                            --------
    num_allocs = atomic_dec(type.num_allocs)
                                                    num_allocs = atomic_inc(type.num_allocs)
                                                    if (num_allocs == 1)
                                                        type.str = str
    if (!num_allocs)
        type.str = NULL
        mem_acct_unref(mem_acct)
                                                        atomic_inc(mem_acct.refcnt)
```

As a general rule, you can't use the result of an atomic operation to do conflicting actions. There are some exceptions under some circumstances, but they don't apply here (at most we could save the ref counting update if the caller always holds another reference, i.e. it's always called by the xlator that has an active ref to mem_acct).

Reply:
I believe memory accounting is useful for debug a leak issue, it does not impact to user. As we have discussed over chat i have deleted typstr to NULL from __gf_free and during first time memory allocation the code save typestr.The only drawback is during last cleanup typestr still available but it is ok because we will handle typestr based on the num_allocs